### PR TITLE
License request/feature

### PR DESCRIPTION
**Pull request by WiseoneO**

### CODE REVIEW

File: src/license-category/license-category.controller.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    async view(
+        @Param('id') id: string
+    ) {
+        return this.licenseCategoryService.view(Number(id))
+    }
+
+    @Post()
+    @UseGuards(AuthGuard())
+    @SetMetadata('action', Actions.Manage)
+    @SetMetadata('subject', Subjects.User)
```
Comment (Chibuzo) [inline src/license-category/license-category.controller.ts:36]: Subject should be admin

File: src/license-category/license-category.controller.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    @ApiOperation({ summary: 'Creates new license category' })
+    async createLicense(
+        @Body() licenseCategoryDto: LicenseCategoryDto
+    ) {
+        return this.licenseCategoryService.create(licenseCategoryDto)
+    }
+
+    @Patch('/:id')
```
Comment (Chibuzo) [inline src/license-category/license-category.controller.ts:44]:
```suggestion
    @Put('/:id')
```

File: src/license-category/license-category.service.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    async getOne(criteria){
+
+        const category = await this.prisma.licenseCategory.findUnique({
+            where: {...criteria}
+        });
+        return category
+    }
+
+    async view(id: Number){
+        const category = await this.getOne({id})
```
Comment (Chibuzo) [inline src/license-category/license-category.service.ts:29]:
```suggestion
    async view(criteria: Record<any, unknown>){
        const category = await this.getOne(criteria)
```

File: src/license-category/license-category.controller.ts (hunk 2 above)
Comment (WiseoneO) [reply src/license-category/license-category.controller.ts:44]: Noted

File: src/license-category/license-category.service.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+        if (!license) {
+            throw new HttpException('License not found', HttpStatus.NOT_FOUND);
+        }
+
+         await this.prisma.licenseCategory.delete({
+            where: {id: id}
+        });
+
+        return {status: 'success', message: 'License category deleted'};
```
Comment (Chibuzo) [inline src/license-category/license-category.service.ts:71]:
```suggestion
         return this.prisma.licenseCategory.delete({
            where: {id: id}
        });
```

File: src/license-request/dto/updateLicenseRequest.dto.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    readonly requestType: LicenceRequestType;
+
+    @IsNotEmpty()
+    @IsNumber()
+    readonly licenseCategoryId: number;
+
+    @IsNotEmpty()
+    @IsOptional()
+    @IsString()
+    readonly paymentId: string;
```
Comment (Chibuzo) [inline src/license-request/dto/updateLicenseRequest.dto.ts:17]:
```suggestion
```
We should change payment id from direct client request

File: src/license-request/license-request.controller.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    @Get('/')
+    @UseGuards(AuthGuard())
+    @SetMetadata('action', Actions.Read)
+    @SetMetadata('subject', Subjects.Company)
+    @ApiOperation({ summary: 'Gets all requests' })
+    async getAll(){
+        return await this.licenseRequestService.getAll();
+    }
+
+    @Get('/company/:companyId/request/:requestId')
```
Comment (Chibuzo) [inline src/license-request/license-request.controller.ts:28]:
```suggestion
    @Get('/company/:companyId/licenserequest/:requestId')
```

File: src/license-request/license-request.controller.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    @SetMetadata('action', Actions.Read)
+    @SetMetadata('subject', Subjects.Company)
+    @ApiOperation({ summary: 'Gets a specific company license request' })
+    @UseInterceptors(DirectorInterceptor)
+    async view(
+        @Param('companyId') companyId: string,
+        @Param('requestId') requestId:string,
+
+    ){
+        return this.licenseRequestService.view( Number(companyId),Number(requestId),)
```
Comment (Chibuzo) [inline src/license-request/license-request.controller.ts:39]: view() should accept an object

File: src/license-request/license-request.controller.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import { ApiOperation } from '@nestjs/swagger';
+import { AuthGuard } from '@nestjs/passport';
+import { CurrentUser } from 'src/auth/decorators/current-user.decorator';
+import { Actions } from 'src/casl/actions.enum';
+import { CanActAuthguard } from 'src/auth/guard/canact.auth.guard';
+import { Subjects } from 'src/casl/subjects.enum';
+import { User } from '@prisma/client';
+
+
+@Controller('license')
```
Comment (Chibuzo) [inline src/license-request/license-request.controller.ts:15]:
```suggestion
@Controller('licenserequest')
```

File: src/license-request/license-request.controller.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    @UseInterceptors(DirectorInterceptor)
+    async view(
+        @Param('companyId') companyId: string,
+        @Param('requestId') requestId:string,
+
+    ){
+        return this.licenseRequestService.view( Number(companyId),Number(requestId),)
+    }
+
+    @Post('/request/company/:companyId')
```
Comment (Chibuzo) [inline src/license-request/license-request.controller.ts:42]:
```suggestion
    @Post('')
```

File: src/license-request/license-request.controller.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    @UseGuards(AuthGuard())
+    @SetMetadata('action', Actions.Create)
+    @SetMetadata('subject', Subjects.Company)
+    @ApiOperation({ summary: 'Creates new license request' })
+    async create(
+        @Param('companyId') companyId: string,
```
Comment (Chibuzo) [inline src/license-request/license-request.controller.ts:48]:
```suggestion
```
companyId should come from the request body

File: src/license-request/license-request.controller.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        @Body() licenseRequestDto: LicenseRequestDto,
+        @CurrentUser() user: User,
+    ){
+        return await this.licenseRequestService.create(Number(companyId),licenseRequestDto)
+    }
+
+    @Delete('/request/:requestId/delete')
```
Comment (Chibuzo) [inline src/license-request/license-request.controller.ts:55]:
```suggestion
    @Delete('/:requestId')
```

File: src/license-request/license-request.controller.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    @UseGuards(AuthGuard())
+    @SetMetadata('action', Actions.Manage)
+    @SetMetadata('subject', Subjects.User)
```
Comment (Chibuzo) [inline src/license-request/license-request.controller.ts:58]:
```suggestion
    @SetMetadata('subject', Subjects.Company)
```

File: src/license-request/license-request.service.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    async view(companyId: any, requestId:Number, ){
+        
+        const company = await this.companyService.getOne({id: companyId})
+        if (!company) throw new HttpException('Company not found', HttpStatus.NOT_FOUND);
+
+        const license = await this.getOne({id: requestId});
+        if (!license) throw new HttpException('License not found', HttpStatus.NOT_FOUND);
+
+
+        return {license};
```
Comment (Chibuzo) [inline src/license-request/license-request.service.ts:37]: Rewrite this function. It should accept an object as argument. Query the LicenseRequest for both companyId and license id at the same time

File: src/license-request/license-request.service.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+    }
+
+    
+
+    async create(companyId: number, licenseRequestDto: LicenseRequestDto){
```
Comment (Chibuzo) [inline src/license-request/license-request.service.ts:43]: companyId should be part of LicenseRequestDto

File: src/license-request/license-request.service.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+        const company =  await this.companyService.getOne({id: companyId})
+        
+        if(!company) throw new HttpException('Compant not found', HttpStatus.NOT_FOUND)
```
Comment (Chibuzo) [inline src/license-request/license-request.service.ts:47]: Use NotFoundException types for these...

File: src/license-request/license-request.service.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+
+        if (!license) {
+            throw new HttpException('License not found', HttpStatus.NOT_FOUND);
+        }
+
+         await this.prisma.licenseRequest.delete({
+            where: {id: requestId}
+        });
+
+        return {status: 'success', message: 'License request deleted'};
```
Comment (Chibuzo) [inline src/license-request/license-request.service.ts:100]: Response interceptor adds status to all responses. You may want to add message to it. However, message should be returned from the controller

**Review by Chibuzo — CHANGES_REQUESTED** [review]
Please effect the suggested changes asap

File: src/license-category/license-category.controller.ts (hunk 1 above)
Comment (Chibuzo) [reply src/license-category/license-category.controller.ts:36]: Add it

**Pull request closed without merging**